lib/good_job/job.rb

```diff
@@ -189,37 +189,27 @@ def perform
       self.performed_at = Time.current
       save! if GoodJob.preserve_job_records
 
-      result, rescued_error = execute
+      result, unhandled_error = execute
 
-      retry_or_discard_error = GoodJob::CurrentExecution.error_on_retry ||
-                               GoodJob::CurrentExecution.error_on_discard
+      result_error = nil
+      result, result_error = nil, result if result.is_a?(Exception) # rubocop:disable Style/ParallelAssignment
 
-      error = nil
-      if rescued_error
-        error = rescued_error
-      elsif result.is_a?(Exception)
-        error = result
-        result = nil
-      elsif retry_or_discard_error
-        error = retry_or_discard_error
-      end
+      job_error = unhandled_error ||
+                  result_error ||
+                  GoodJob::CurrentExecution.error_on_retry ||
+                  GoodJob::CurrentExecution.error_on_discard
 
-      error_message = "#{error.class}: #{error.message}" if error
-      self.error = error_message
+      self.error = "#{job_error.class}: #{job_error.message}" if job_error
 
-      if rescued_error && GoodJob.reperform_jobs_on_standard_error
+      if unhandled_error && GoodJob.reperform_jobs_on_standard_error
         save!
+      elsif GoodJob.preserve_job_records == true || (GoodJob.preserve_job_records == :on_unhandled_error && unhandled_error)
+        update!(finished_at: Time.current)
```

Comment:
Style: If a model object has been previously dirtied, `save` should be used rather than `update`.